src/block/Furnace.php

```diff
@@ -81,10 +146,121 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
+			$this->position->getWorld()->setBlock($this->getPosition(), $this);
+		}
+	}
+
+	protected function onStopSmelting() : void{
+		if($this->isLit()){
+			$this->setLit(false);
+			$this->remainingFuelTime = $this->cookTime = $this->maxFuelTime = 0;
+			$this->position->getWorld()->setBlock($this->getPosition(), $this);
+		}
+	}
+
+	protected function checkFuel(Item $fuel) : void{
+		$tile = $this->position->getWorld()->getTile($this->position);
+		if($tile instanceof TileFurnace && !$tile->isClosed()){
+			$ev = new FurnaceBurnEvent($tile, $fuel, $fuel->getFuelTime());
+			$ev->call();
+			if($ev->isCancelled()){
+				return;
+			}
+
+			$this->maxFuelTime = $this->remainingFuelTime = $ev->getBurnTime();
+			$this->onStartSmelting();
+
+			if($this->remainingFuelTime > 0 && $ev->isBurning()){
+				$tile->getInventory()->setFuel($fuel->getFuelResidue());
+			}
+		}
+	}
+
 	public function onScheduledUpdate() : void{
-		$furnace = $this->position->getWorld()->getTile($this->position);
-		if($furnace instanceof TileFurnace and $furnace->onUpdate()){
-			$this->position->getWorld()->scheduleDelayedBlockUpdate($this->position, 1); //TODO: check this
+		$tile = $this->position->getWorld()->getTile($this->position);
+		if($tile instanceof TileFurnace && !$tile->isClosed()){
+			$inventory = $tile->getInventory();
+
+			$prevCookTime = $this->cookTime;
+			$prevRemainingFuelTime = $this->remainingFuelTime;
+			$prevMaxFuelTime = $this->maxFuelTime;
+
+			$fuel = $inventory->getFuel();
+			$raw = $inventory->getSmelting();
+			$product = $inventory->getResult();
+
+			$furnaceType = $tile->getFurnaceType();
+			$smelt = $this->position->getWorld()->getServer()->getCraftingManager()->getFurnaceRecipeManager($furnaceType)->match($raw);
+			$canSmelt = ($smelt instanceof FurnaceRecipe and $raw->getCount() > 0 and (($smelt->getResult()->equals($product) and $product->getCount() < $product->getMaxStackSize()) or $product->isNull()));
+
+			if($this->remainingFuelTime <= 0 and $canSmelt and $fuel->getFuelTime() > 0 and $fuel->getCount() > 0){
+				$this->checkFuel($fuel);
+			}
+
+			if($this->remainingFuelTime > 0){
+				--$this->remainingFuelTime;
+
+				if($smelt instanceof FurnaceRecipe and $canSmelt){
+					++$this->cookTime;
+
+					if($this->cookTime >= $furnaceType->getCookDurationTicks()){
+						$product = $smelt->getResult()->setCount($product->getCount() + 1);
+
+						$ev = new FurnaceSmeltEvent($tile, $raw, $product);
+						$ev->call();
+
+						if(!$ev->isCancelled()){
+							$inventory->setResult($ev->getResult());
+							$raw->pop();
+							$inventory->setSmelting($raw);
+						}
+
+						$this->cookTime -= $furnaceType->getCookDurationTicks();
+					}
+					$this->position->getWorld()->setBlock($this->position, $this);
+				}elseif($this->remainingFuelTime <= 0){
+					$this->onStopSmelting();
+				}else{
+					$this->cookTime = 0;
+					$this->position->getWorld()->setBlock($this->position, $this);
+				}
+				if(mt_rand(1, 100) < 5){
```

Comment:
Where did you get these values from?
I currently haven't looked into IDA, but there is no information about the probability in the wiki. 
(I did look into Java and it is a probability of 10% there instead of 4%.)
```java
if (random.nextDouble() < 0.1D) {...}
```

Reply:
Also, in Java the position of the sound would be:
```php
new Vector3(
    $this->position->x + 0.5,
    $this->position->y,
    $this->position->z + 0.5
);
```
But that also would need to be checked with IDA. Just wanted to mention, that those values in the PR could not be true.

Reply:
> Where did you get these values from? I currently haven't looked into IDA, but there is no information about the probability in the wiki. (I did look into Java and it is a probability of 10% there instead of 4%.)
> 
> ```java
> if (random.nextDouble() < 0.1D) {...}
> ```

I have actually set 4% for the random sound because after many tests, it is the one that seemed closer to vanilla. No precise official value is put on the wiki, we only note "Plays randomly when active". 10% seems a lot, the sound is present all the time, while it is not necessarily present all the time in vanilla